text/0059-no-domain-upwards-propagation.md

```diff
@@ -0,0 +1,73 @@
+- Start Date: 2024-03-18
+- RFC PR: [amaranth-lang/rfcs#59](https://github.com/amaranth-lang/rfcs/pull/59)
+- Amaranth Issue: [amaranth-lang/amaranth#0000](https://github.com/amaranth-lang/amaranth/issues/0000)
+
+# Get rid of upwards propagation of clock domains
+
+## Summary
+[summary]: #summary
+
+Upwards propagation of clock domains is removed, `local=True` effectively becomes the default.
+
+## Motivation
+[motivation]: #motivation
+
+Currently, a clock domain created anywhere in the hierarchy is, by default, propagated everywhere within that hierarchy. The other option is marking a clock domain with `local=True`, which restricts propagation to only downwards from the point of definition.
+
+The default behavior constitutes the ultimate implicit spooky action at a distance, making it hard to reason about clock domain origins in complex Amaranth design. It is also rarely used. This dangerous default should be changed to something more reasonable.
+
+## Guide and reference-level explanation
+[explanation]: #explanation
+
+In version 0.5, upwards domain propagation becomes deprecated. Any use of a domain that wouldn't be valid without upwards propagation (ie. wouldn't be valid if the domain were `local=True`) triggers a deprecation warning.
+
+In version 0.6, upwards domain propagation is removed entirely. The `local` argument to `ClockDomain` constructor becomes redundant and is deprecated.
+
+In version 0.7, the `local` argument is removed.
+
+For the cases where upwards domain propagation was used intentionally, a new mechanism is introduced to explicitly import a clock domain from a submodule:
```

Comment:
I think this is completely unnecessary. We should not be adding more `getattr`-like functionality where we can avoid it.

Is there a value-add over `m.domains.sync = clkgen.cd_sync`?

Reply:
Slight but non-zero: with this proposal, `clkgen` can defer creating `cd_sync` until `elaborate`, where it can decide things (including polarity) based on the platform

Reply:
Right. I would expect clock/reset generators to be so platform-specific in first place that the clock/reset generator itself would be created in `elaborate`, so this advantage is fairly slim.